.evergreen/run-socks5-tests.sh

```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
+
+set -o errexit  # Exit the script with error if any of the commands fail
+set -o xtrace  # For debuggability, no external credentials are used here
+
+source "${PROJECT_DIRECTORY}/.evergreen/init-nvm.sh"
+
+PYTHON_BINARY=${PYTHON_BINARY:-python3}
+
+# ssl setup
+SSL=${SSL:-nossl}
+if [ "$SSL" != "nossl" ]; then
+   export SSL_KEY_FILE="$DRIVERS_TOOLS/.evergreen/x509gen/client.pem"
+   export SSL_CA_FILE="$DRIVERS_TOOLS/.evergreen/x509gen/ca.pem"
+fi
+
+# Grab a connection string that only refers to *one* of the hosts in MONGODB_URI
+FIRST_HOST=$(node -p 'new (require("mongodb-connection-string-url").default)(process.env.MONGODB_URI).hosts[0]')
+# Use localhost:12345 as the URL for the single host that we connect to,
+# we configure the Socks5 proxy server script to redirect from this to FIRST_HOST
+export MONGODB_URI_SINGLEHOST="mongodb://localhost:12345/"
+
+# Compute path to socks5 fake server script in a way that works on Windows
+SOCKS5_SERVER_SCRIPT="${PROJECT_DIRECTORY}/.evergreen/socks5srv.py"
+if [ "Windows_NT" = "$OS" ]; then
+  SOCKS5_SERVER_SCRIPT=$(cygpath -w "$SOCKS5_SERVER_SCRIPT")
+fi
+
+# First, test with Socks5 + authentication required
+"$PYTHON_BINARY" "$SOCKS5_SERVER_SCRIPT" --port 1080 --auth username:p4ssw0rd --map "localhost:12345 to $FIRST_HOST" &
+PID=$!
+env SOCKS5_CONFIG='["localhost",1080,"username","p4ssw0rd"]' npm run check:socks5
+kill $PID
+
+# Second, test with Socks5 + no authentication
+"$PYTHON_BINARY" "$SOCKS5_SERVER_SCRIPT" --port 1081 --map "localhost:12345 to $FIRST_HOST" &
+PID=$!
+env SOCKS5_CONFIG='["localhost",1081]' npm run check:socks5
+kill $PID
+
```

Comment:
I think we could add `npx mocha --file test/tools/runner/index.js test/functional/client_side_encryption` here and check that FLE is working over the proxy. I think you'll need to pull in a few more variables in the EVG config, you can take a look at the top of run-custom-csfle script.

We could also totally make this follow up work, would it be possible to put together a `MONGODB_URI` variable that would just "work" with the existing tests? (They shouldn't have any special requirements for the URI)

Reply:
Yeah, that’s a great idea. I’ve added an invocation of `bash "${PROJECT_DIRECTORY}/.evergreen/run-custom-csfle-tests.sh"` here and modified the test runner so that it forwards proxy options in its configuration where appropriate, please take a look and let me know what I messed up :slightly_smiling_face:

Reply:
@nbbeeken Actually, it looks like some more options forwarding is needed in order for the TLS version of these tests to pass … is that worth it or should the CSFLE tests only be run when the server does not require TLS? I’m happy to go either way.

Reply:
I think you can gate it on running only when TLS is disabled. The important part to me of checking FLE over the proxy is that when we create an internal separate client, that it is handed the same config options and works as expected. The FLE deadlock tests will set up and run the scenario where a separate MongoClient instance is needed, and it _should_ inherit everything but that's what I want to see pass.

Reply:
@nbbeeken Sure, done – tests pass :) I noticed that the xunit.xml files from different mocha runs overwrite each other, not sure what to do about that but it also feels like more of a cosmetic issue (unless you feel strongly about it).

Reply:
Actually, it doesn’t pass on Windows either: 

```
[2021/12/03 11:09:25.960] + '[' nossl == nossl ']'
[2021/12/03 11:09:25.960] + env 'MONGODB_URI=mongodb://localhost:12345/?proxyHost=localhost&proxyUsername=username&proxyPassword=p4ssw0rd' bash /cygdrive/c/data/mci/72615ac373d7b45a5862ded1dd3d86af/src/.evergreen/run-custom-csfle-tests.sh
[2021/12/03 11:09:26.009] + set -o errexit
[2021/12/03 11:09:26.009] + echo 'adding temporary AWS credentials to environment'
[2021/12/03 11:09:26.009] + . Z:/data/mci/72615ac373d7b45a5862ded1dd3d86af/drivers-tools/.evergreen/csfle/set-temp-creds.sh
[2021/12/03 11:09:26.069] ++ set +o xtrace
[2021/12/03 11:09:26.069] Traceback (most recent call last):
[2021/12/03 11:09:26.069]   File "<stdin>", line 1, in <module>
[2021/12/03 11:09:26.069] ImportError: No module named boto3
[2021/12/03 11:26:37.398] Command stopped early: context canceled
```

Not sure what’s up with that, but I assume it’s fine because right now the csfle tests only run on Ubuntu to begin with?

Reply:
Yea we've decided its enough coverage to just try FLE on ubuntu for now. So we could gate running the fle script to only linux. And yea its fine that the xunit gets over written, we'll still see failures so I agree mostly cosmetic.